landscape/lib/network.py

```diff
@@ -10,129 +10,63 @@
 
-# from header /usr/include/bits/ioctls.h
-SIOCGIFCONF = 0x8912
-SIOCGIFFLAGS = 0x8913
-SIOCGIFNETMASK = 0x891b
-SIOCGIFBRDADDR = 0x8919
-SIOCGIFADDR = 0x8915
-SIOCGIFHWADDR = 0x8927
 
+SIOCGIFFLAGS = 0x8913  # from header /usr/include/bits/ioctls.h
 SIOCETHTOOL = 0x8946  # As defined in include/uapi/linux/sockios.h
 ETHTOOL_GSET = 0x00000001  # Get status command.
 
-# struct definition from header /usr/include/net/if.h
-# the struct size varies according to the platform arch size
-# a minimal c program was used to determine the size of the
-# struct, standard headers removed for brevity.
-"""
-#include <linux/if.h>
-int main() {
-  printf("Size of struct %lu\n", sizeof(struct ifreq));
-}
-"""
-
-IF_STRUCT_SIZE_32 = 32
-IF_STRUCT_SIZE_64 = 40
-
 
 def is_64():
     """Returns C{True} if the platform is 64-bit, otherwise C{False}."""
     return struct.calcsize("l") == 8
 
 
-# initialize the struct size as per the machine's architecture
-IF_STRUCT_SIZE = is_64() and IF_STRUCT_SIZE_64 or IF_STRUCT_SIZE_32
-
-
-def get_active_interfaces(sock):
-    """Generator yields active network interface names.
-
-    @param sock: a socket instance.
-    """
-    max_interfaces = 128
+def get_active_interfaces():
+    for interface in netifaces.interfaces():
+        ifaddresses = netifaces.ifaddresses(interface)
+        # Skip interfaces with no IPv4 addresses.
+        inet_addr = ifaddresses.get(netifaces.AF_INET, [{}])[0].get('addr')
+        if inet_addr:
+            yield interface, ifaddresses
 
-    # Setup an array to hold our response, and initialized to null strings.
-    interfaces = array.array("B", b"\0" * max_interfaces * IF_STRUCT_SIZE)
-    buffer_size = interfaces.buffer_info()[0]
-    packed_bytes = struct.pack(
-        "iL", max_interfaces * IF_STRUCT_SIZE, buffer_size)
 
-    byte_length = struct.unpack(
-        "iL", fcntl.ioctl(sock.fileno(), SIOCGIFCONF, packed_bytes))[0]
-
-    result = interfaces.tostring()
-
-    # Generator over the interface names
-    already_found = set()
-    for index in range(0, byte_length, IF_STRUCT_SIZE):
-        ifreq_struct = result[index:index + IF_STRUCT_SIZE]
-        interface_name = ifreq_struct[:ifreq_struct.index(b"\0")]
-        if interface_name not in already_found:
-            already_found.add(interface_name)
-            yield interface_name
-
-
-def get_broadcast_address(sock, interface):
+def get_broadcast_address(ifaddresses):
     """Return the broadcast address associated to an interface.
 
-    @param sock: a socket instance.
     @param interface: The name of the interface.
     """
-    return socket.inet_ntoa(fcntl.ioctl(
-        sock.fileno(),
-        SIOCGIFBRDADDR,
-        struct.pack("256s", interface[:15]))[20:24])
+    return ifaddresses[netifaces.AF_INET][0].get('broadcast', '0.0.0.0')
 
 
-def get_netmask(sock, interface):
+def get_netmask(ifaddresses):
     """Return the network mask associated to an interface.
```

Comment:
Docstring needs updates

Reply:
Docstrings updated in all affected `get_*` functions.